fastval.go

```diff
@@ -166,107 +217,196 @@ func (val FastVal) GetTime() *time.Time {
+		return parsedVal, err == nil
 	case JsonUintValue:
-		parsedVal, _ := strconv.ParseUint(string(val.sliceData), 10, 64)
-		return parsedVal
+		parsedVal, err := strconv.ParseUint(string(val.sliceData), 10, 64)
+		return parsedVal, err == nil
 	case JsonFloatValue:
-		parsedVal, _ := strconv.ParseFloat(string(val.sliceData), 64)
-		return uint64(parsedVal)
+		parsedVal, err := strconv.ParseFloat(string(val.sliceData), 64)
+		return uint64(parsedVal), err == nil
 	case TrueValue:
-		return 1
+		return 1, true
 	case FalseValue:
-		return 0
-	case NullValue:
-		return math.MaxUint64
+		return 0, true
 	}
-	return 0
+	return 0, false
 }
 
-func (val FastVal) AsFloat() float64 {
+func (val FastVal) AsFloat() (float64, bool) {
 	switch val.dataType {
 	case IntValue:
-		return float64(val.GetInt())
+		return float64(val.GetInt()), true
 	case UintValue:
-		return float64(val.GetUint())
+		return float64(val.GetUint()), true
 	case FloatValue:
-		return val.GetFloat()
+		return val.GetFloat(), true
 	case JsonIntValue:
-		parsedVal, _ := strconv.ParseInt(string(val.sliceData), 10, 64)
-		return float64(parsedVal)
+		parsedVal, err := strconv.ParseInt(string(val.sliceData), 10, 64)
+		return float64(parsedVal), err == nil
 	case JsonUintValue:
-		parsedVal, _ := strconv.ParseUint(string(val.sliceData), 10, 64)
-		return float64(parsedVal)
+		parsedVal, err := strconv.ParseUint(string(val.sliceData), 10, 64)
+		return float64(parsedVal), err == nil
+	case JsonStringValue:
+		fallthrough
 	case JsonFloatValue:
-		parsedVal, _ := strconv.ParseFloat(string(val.sliceData), 64)
-		return parsedVal
+		parsedVal, err := strconv.ParseFloat(string(val.sliceData), 64)
+		return parsedVal, err == nil
 	case TrueValue:
-		return 1.0
+		return 1.0, true
 	case FalseValue:
-		return 0.0
-	case NullValue:
-		return math.MaxFloat64
+		return 0.0, true
 	}
-	return 0.0
+	return 0.0, false
 }
 
-func (val FastVal) AsBoolean() bool {
-	return val.AsInt() != 0 && val.AsInt() != math.MinInt64
+func (val FastVal) AsBoolean() (bool, bool) {
+	switch val.dataType {
+	case JsonStringValue:
+		if TrueStringRegex.Match(val.sliceData) {
+			return true, true
+		} else if FalseStringRegex.Match(val.sliceData) {
+			return false, true
+		} else {
+			return false, false
+		}
+	case IntValue:
+		return val.GetInt() != 0, true
+	case UintValue:
+		return val.GetUint() != 0, true
+	case FloatValue:
+		return val.GetFloat() != 0.0, true
+	case JsonIntValue:
+		parsedVal, err := strconv.ParseInt(string(val.sliceData), 10, 64)
+		return parsedVal != 0, err == nil
+	case JsonUintValue:
+		parsedVal, err := strconv.ParseUint(string(val.sliceData), 10, 64)
+		return parsedVal != 0, err == nil
+	case JsonFloatValue:
+		parsedVal, err := strconv.ParseFloat(string(val.sliceData), 64)
+		return parsedVal != 0.0, err == nil
+	case TrueValue:
+		return true, true
+	case FalseValue:
+		return false, true
+	default:
+		// Undefined
+		return true, false
```

Comment:
isn't it that undefined is treated as false? This is my understanding from the design doc.

Reply:
At this point it doesn't really matter - the parent callers really only care about the second returned bool, the false in this line.